### PR TITLE
Correct deps throughout

### LINKED ISSUE #3068: Document need to install futures on py27
futures (`conda install futures` or `pip install futures`) is needed by the new bokeh server, but isn't caught by `devdeps.py` and also not listed under the install dependencies in the docs (http://bokeh.pydata.org/en/latest/docs/installation.html#install-dependencies)

Not sure where else people might look for the reference.

### LINKED ISSUE #3262: Check dependencies list for release
At a minimum, Flask and pyzmq can be removed as runtime dependencies.

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #3068 closes #3262

This PR removes pandas as a hard dependency (marking it `optional` in the core library and `required` in the subcomponents and sample data that require it)

It updates the conda recipe dependency list and the documentation to match.

### CODE REVIEW

File: bokeh/_json_encoder.py [hunk 1]
```diff
@@ -6,20 +6,12 @@
...
-except ImportError:
-    is_pandas = False
-
-try:
-    from dateutil.relativedelta import relativedelta
-    is_dateutil = True
-except ImportError:
-    is_dateutil = False
+pd = optional('pandas')
+dateutil = optional('dateutil')
```
Comment (bryevdv) [inline bokeh/_json_encoder.py]: It looks like we require `dateutil` in `properties.py` can this just be converted to a regular import?

**Comment by mattpap** [comment]
@bryevdv, would you mind renaming `optional()` to `import_optional()`?

**Comment by bryevdv** [comment]
@mattpap sure tho I think I might prefer `optional_import`. The same for `required` -> `required_import`?

**Comment by bryevdv** [comment]
@mattpap OK to merge?

**Comment by bryevdv** [comment]
@bokeh/core I plan to merge this tomorrow morning unless asked otherwise. To recap:

* remove pandas as a hard dependency
* warn when pandas-using features are used
* remove other out of date dependencies (flask. pyzmq, etc)

**Comment by mattpap** [comment]
LGTM

**Pull request merged**